homeassistant/components/recorder/util.py

```diff
@@ -273,3 +283,25 @@ def end_incomplete_runs(session, start_time):
             "Ended unfinished session (id=%s from %s)", run.run_id, run.start
         )
         session.add(run)
+
+
+def try_database_job(
+    instance: Recorder, description: str, job: callable, *args, **kwargs
+):
+    """Try to execute a database job."""
+    try:
+        return job(*args, **kwargs)
+    except OperationalError as err:
+        if (
+            instance.engine.dialect.name == "mysql"
+            and err.orig.args[0] in RETRYABLE_MYSQL_ERRORS
+        ):
+            _LOGGER.info(
+                "%s; %s not completed, retrying", err.orig.args[1], description
+            )
+            time.sleep(instance.db_retry_wait)
+            return False
+
+        _LOGGER.warning("Error executing %s: %s", description, err)
+
+    return False
```

Comment:
Should we return `True` here because the error is not retriable?

Reply:
If we need to know the difference between a perm failure and a temp failure we change the return type to an enum or similar.

Reply:
The old code returned `True` both if the purge succeeded and if it failed with a permanent error, I'll revert to that.